8320362: Load anchor certificates from Keychain keystore

alexeybakhtin opened the pull request:
Please review the proposed fix.

The patch loads system root certificates from the MacOS Keychain with TrustSettings.
It allows to build a trusted certificate path using the MacOS Keychain store only.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8321045](https://bugs.openjdk.org/browse/JDK-8321045) to be approved

### Issues
 * [JDK-8320362](https://bugs.openjdk.org/browse/JDK-8320362): Load anchor certificates from Keychain keystore (**Enhancement** - P4)
 * [JDK-8321045](https://bugs.openjdk.org/browse/JDK-8321045): Load anchor certificates from Keychain keystore (**CSR**)


### Reviewers
 * @1keep2keepFaith (no known openjdk.org user name / role) ⚠️ Review applies to [db98b7c2](https://git.openjdk.org/jdk/pull/16722/files/db98b7c2ed0250edb13bf54167d207f57ab7ac2a)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to [24ca1206](https://git.openjdk.org/jdk/pull/16722/files/24ca1206b3ba1d5dfdbab2e1042108b6543a4283)
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**) ⚠️ Review applies to [b92e21b9](https://git.openjdk.org/jdk/pull/16722/files/b92e21b97c680e77e76b86336200631a641cdbfb)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16722/head:pull/16722` \
`$ git checkout pull/16722`

Update a local copy of the PR: \
`$ git checkout pull/16722` \
`$ git pull https://git.openjdk.org/jdk.git pull/16722/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16722`

View PR using the GUI difftool: \
`$ git pr show -t 16722`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16722.diff">https://git.openjdk.org/jdk/pull/16722.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16722#issuecomment-1818342769)

1keep2keepFaith commented:
Thanks please let me know if I need to do other

On Fri, Dec 1, 2023, 7:04 PM Alexey Bakhtin ***@***.***>
wrote:

> Please review the proposed fix.
>
> The patch loads system root certificates from the MacOS Keychain with
> TrustSettings.
> It allows to build a trusted certificate path using the MacOS Keychain
> store only.
> ------------------------------
> Progress
>
>    - Change must be properly reviewed (1 review required, with at least 1
>    Reviewer <https://openjdk.org/bylaws#reviewer>)
>    - Change must not contain extraneous whitespace
>    - Commit message must refer to an issue
>    - Change requires a CSR request matching fixVersion 22 to be approved
>    (needs to be created)
>
> Issue
>
>    - JDK-8320362 <https://bugs.openjdk.org/browse/JDK-8320362>: Load
>    anchor certificates from Keychain keystore (*Enhancement* - P4)
>
> Reviewers
>
>    - @1keep2keepFaith <https://github.com/1keep2keepFaith> (no known
>    openjdk.org user name / role)
>
> Reviewing Using git
>
> Checkout this PR locally:
> $ git fetch https://git.openjdk.org/jdk.git pull/16722/head:pull/16722
> $ git checkout pull/16722
>
> Update a local copy of the PR:
> $ git checkout pull/16722
> $ git pull https://git.openjdk.org/jdk.git pull/16722/head
> Using Skara CLI tools
>
> Checkout this PR locally:
> $ git pr checkout 16722
>
> View PR using the GUI difftool:
> $ git pr show -t 16722
> Using diff file
>
> Download this PR as a diff file:
> <https://git.openjdk.org/jdk/pull/16722.diff>
> https://git.openjdk.org/jdk/pull/16722.diff
> Webrev
>
> Link to Webrev Comment
> <https://git.openjdk.org/jdk/pull/16722#issuecomment-1818342769>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/openjdk/jdk/pull/16722
> Commit Summary
>
>    - b0d9179
>    <https://github.com/openjdk/jdk/pull/16722/commits/b0d91794f9cae4e9f60138aae1c5b9ecb743f128>
>    8320362: Load anchor certificates from Keychain keystore
>    - db98b7c
>    <https://github.com/openjdk/jdk/pull/16722/commits/db98b7c2ed0250edb13bf54167d207f57ab7ac2a>
>    Mark trust anchor certificates readonly
>
> File Changes
>
> (2 files <https://github.com/openjdk/jdk/pull/16722/files>)
>
>    - *M* src/java.base/macosx/classes/apple/security/KeychainStore.java
>    <https://github.com/openjdk/jdk/pull/16722/files#diff-3671bea052d656bf293d190536c9b3d0088e4b0ec9ef2b27bf05c8e8fb56de10>
>    (19)
>    - *M* src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m
>    <https://github.com/openjdk/jdk/pull/16722/files#diff-b0b2f959f0fa18e568c0688bd233882283bca94f4ff6dce886a21510ff82d64d>
>    (102)
>
> Patch Links:
>
>    - https://github.com/openjdk/jdk/pull/16722.patch
>    - https://github.com/openjdk/jdk/pull/16722.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openjdk/jdk/pull/16722>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BEGMDPGVFRLGCSIUBCE3OS3YHJ5B3AVCNFSM6AAAAAA7QUNTROVHI2DSMVQWIX3LMV43ASLTON2WKOZSGAYDAMJXGMZTANY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

alexeybakhtin commented:
As suggested by @wangweij, the new Keychain-ROOT keystore is introduced for the trusted anchor certificates.
The Keychain-ROOT keystore is read-only and throws KeyStoreException in an attempt to modification

alexeybakhtin commented:
> What are the change for existing `addCertificatesToKeystore` for? Is there any behavior change?

Hi @wangweij . No behavior changes. Just reformatted to make it similar to addCertificatesToKeystoreRoot. Can be reverted back.

alexeybakhtin commented on src/java.base/macosx/classes/apple/security/AppleProvider.java:
```diff
@@ -82,7 +84,9 @@ public AppleProvider() {
-                           "KeychainStore", "apple.security.KeychainStore"));
+                           "KeychainStore", "apple.security.KeychainStore.USER"));
+                putService(new ProviderService(p, "KeyStore",
+                           "KeychainStore-ROOT", "apple.security.KeychainStore.ROOT"));
```
Thank you. updated with the $ sign

alexeybakhtin commented on src/java.base/macosx/classes/apple/security/KeychainStore.java:
```diff
@@ -42,12 +42,72 @@
  */
 
-public final class KeychainStore extends KeyStoreSpi {
+abstract class KeychainStore extends KeyStoreSpi {
```
Good catch. Thank you. Added `sealed`

alexeybakhtin commented on src/java.base/macosx/classes/apple/security/KeychainStore.java:
```diff
@@ -137,14 +197,25 @@ private static void permissionCheck() {
+     * Returns the name of the keystore.
+     */
+    private String getName()
+    {
```
OK. Not much reason for this method. removed

alexeybakhtin commented on src/java.base/macosx/classes/apple/security/KeychainStore.java:
```diff
@@ -472,6 +543,7 @@ public void engineSetKeyEntry(String alias, byte[] key,
         permissionCheck();
 
         synchronized(entries) {
+
```
Fixed

alexeybakhtin commented on src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:
```diff
@@ -471,32 +471,21 @@ static void addCertificatesToKeystore(JNIEnv *env, jobject keyStore)
+                trustSettings = NULL;
+                // Load user trustSettings into inputTrust
+                if (SecTrustSettingsCopyTrustSettings(certRef, domain, &trustSettings) == errSecSuccess && trustSettings != NULL) {
+                    if(inputTrust == NULL) {
```
Fixed

alexeybakhtin commented on src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:
```diff
@@ -519,6 +508,102 @@ static void addCertificatesToKeystore(JNIEnv *env, jobject keyStore)
+    }
+
+    // Read Trust Anchors
+    if(SecTrustCopyAnchorCertificates(&currAnchors) == errSecSuccess) {
```
Fixed

alexeybakhtin commented:
> Is it possible to reuse some some lines from `addCertificatesToKeystore`?
> 
> BTW, I reviewed the CSR.

Hi @wangweij,
Thank you a lot for PR and CSR review. I have updated PR with review findings and refactored addCertificatesToKeystore/addCertificatesToKeystoreRoot to reuse some code

alexeybakhtin commented:
> Do you intend to call `addIdentitiesToKeystore` for both store types?

Yes, you are right. Thank you. We do not need private identities in the root keystore. Updated

alexeybakhtin commented:
> Great! The change looks good. Can you manage to add a test? Maybe try to load both store types. Make sure they have different contents and not empty (?).

Thank you. I have updated `java/security/KeyStore/CheckMacOSKeyChainTrust.java` test to load system root certificates and compare it with the content of the `KEYCHAINSTORE-ROOT` keystore.

alexeybakhtin commented on test/jdk/java/security/KeyStore/CheckMacOSKeyChainTrust.java:
```diff
@@ -32,22 +32,32 @@
+        } else {
+            // check user and admin trustsettings to find distrusted certs
+            loadUser(false);
+            loadAdmin(false);
```
Yes. Exactly. The trusted cert can be distrusted in the user/admin domain, so It should not be available in the KEYCHAINSTORE-ROOT

alexeybakhtin commented:
> Will KEYCHAINSTORE-ROOT contains trusted certs inside KEYCHAINSTORE? I tried on my machine and there are some items not in `security dump-trust-settings -s`.

I've found an issue in the original implementation that prevents adding different certificates with the same CN. I am going to fix it as part of this PR because it affects the scope of root certificates in the KEYCHAINSTORE-ROOT

alexeybakhtin commented:
> Will KEYCHAINSTORE-ROOT contains trusted certs inside KEYCHAINSTORE? I tried on my machine and there are some items not in `security dump-trust-settings -s`.
`security dump-trust-settings -s` returns only predefined root certificates. KEYCHAINSTORE-ROOT additionally contains installed root trusted certificates in the system domain

alexeybakhtin commented on test/jdk/java/security/KeyStore/CheckMacOSKeyChainTrust.java:
```diff
@@ -32,22 +32,32 @@
  * @summary Check whether loading of certificates from MacOS Keychain correctly
  *          honors trust settings
+ * @run main CheckMacOSKeyChainTrust KEYCHAINSTORE
+ * @run main CheckMacOSKeyChainTrust KEYCHAINSTORE-ROOT
```
Yes, we can, but it looks like not required until we compare the contents of these stores

alexeybakhtin commented:
> > `security dump-trust-settings -s` returns only predefined root certificates. KEYCHAINSTORE-ROOT additionally contains installed root trusted certificates in the system domain
> 
> Are you sure they should be added into this keystore? It looks like all the extra certs in KEYCHAINSTORE-ROOT that are not in `security dump-trust-settings -s` are all inside KEYCHAINSTORE. Maybe that's where they should belong to?

Thank you. You are right. It is better if KEYCHAINSTORE-ROOT contains only predefined roots. Unfortunately, SecTrustCopyAnchorCertificates can not be used in this case.

I have updated the patch to read certificates from the "/System/Library/Keychains/SystemRootCertificates.keychain" keychain

alexeybakhtin commented:
@wangweij could you please take a look at the last proposed commit, if you have any chance

alexeybakhtin commented on src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:
```diff
@@ -515,10 +515,32 @@ static void addCertificatesToKeystoreRoot(JNIEnv *env, jobject keyStore,
-    if (SecTrustCopyAnchorCertificates(&currAnchors) == errSecSuccess) {
+    // Load predefined root certificates from SystemRootCertificates keychain
+    // SecTrustCopyAnchorCertificates includes extra root certificates and can not be used here
+    if( SecKeychainOpen("/System/Library/Keychains/SystemRootCertificates.keychain", &keychain) != errSecSuccess ) {
```
Hi @wangweij , 
Thank you for review.
Unfortunately `SecTrustCopyCustomAnchorCertificates` can not be used also. It is used to retrieve certificates from your own created SecTrust. As I know it is not possible to create/load SecTrust with predefined certs without reading`/System/Library/Keychains/SystemRootCertificates.keychain`

alexeybakhtin replied:
Thank you. I've updated CSR and waiting for review.

alexeybakhtin replied:
CSR is in the Proposed state now. Workflow does not allow me to Finalize it.

alexeybakhtin replied:
Thank you. Moved to Finalized

seanjmullan commented:
Is it practical to add a test as described in the bug?:

`java -Djavax.net.ssl.trustStoreType=KeychainStore-ROOT HttpsURLConnectionTest https://github.com/`

This could be added to the `test/jdk/security/infra` test directory since it depends on an external server.

alexeybakhtin commented:
Hi @seanjmullan

seanjmullan commented:
> Hi @seanjmullan Thank you for review I've added the test

Thanks. @rhalade Is this an acceptable place (security/infra) to put a test that makes external network connections?

rhalade commented:
> > Hi @seanjmullan Thank you for review I've added the test
> 
> Thanks. @rhalade Is this an acceptable place (security/infra) to put a test that makes external network connections?

Yes, it is correct place to add this infra test. @alexeybakhtin please update https://github.com/openjdk/jdk/blob/master/test/jdk/TEST.groups#L254 definition to include this test. Update could be:

jdk_security_infra = \
    security/infra

seanjmullan reviewed (approved):
I only reviewed HttpsURLConnectionTest.java and had one minor comment (typo).

alexeybakhtin commented on test/jdk/security/infra/javax/net/ssl/HttpsURLConnectionTest.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * @test
+ * @bug 8320362
+ * @summary Verifies successfull connection to external server with
```
Thank you. Fixed

alexeybakhtin commented:
/integrate
